doc/SAIP4Extensions/SAI_P4EXT_API_Proposal.md

```diff
@@ -0,0 +1,300 @@
+    sai_set_p4ext_entry_attribute_fn              set_p4ext_entry_attribute;
+    sai_get_p4ext_entry_attribute_fn              get_p4ext_entry_attribute;
+} sai_p4ext_api_t;
+
+```
+
+# API Usage #
+
+Consider a scenario where a Vendor A wishes to extend the SAI pipeline functionality with a custom feature which can be modeled with a P4 match-action table but is not yet supported in SAI. Let say this new feature classifies incoming packets based on packet field tuples SIP,DIP and sets the traffic class tc. As mentioned earlier the proposal does not deal with the implementation details of data pipeline by each individual vendor, thus the relative position of this table in the SAI pipeline is not part of the specification and is vendor dependent. Its upto the vendor to decide whether this table will come after the SAI QoS tables or before and resolve any conflict that would arise from the actions of the QoS tables and the new P4 Extension Table. This simple feature can be expressed in P4 as below.
+
+```
+table flow_classification {
+    action set_tc(switch_tc_t tc) {
+        ig_md.qos.etrap_tc = tc;
+    }
+
+    key = {
+        ig_md.lkp.ip_src_addr[31:0] : ternary @name("src_addr");
+        ig_md.lkp.ip_dst_addr[31:0] : ternary @name("dst_addr");
+    }
+    actions = {
+        set_tc;
+    }
+}
+```
+
+The above P4 table - *flow_classification*, has two key/ match fields.
+1. *src_addr* - This is the incoming packet 32 bit Source IP Address
+2. *dst_addr* - This is the incoming packet 32 bit Destination IP Address
+The type of match for both these fields is ternary. In general the match_kind is expected to be one of the match_kinds
+as defined by the P4 core library. More information please refer to the [P4 16
+Specification](https://p4.org/p4-spec/docs/P4-16-v1.0.0-spec.html#sec-match-kind-type)
+In addition the above table describes just one possible action - *set_tc*. An incoming packet that matches an entry in
+the above table can be programmed to execute this action. The action *set_tc* accepts only one parameter *tc* which is used
+to sed etrap_tc metadata in the P4 data pipeline
+
+## Creating a P4 Extension Table Entry ##
+
+```cpp
+#define P4_TABLE_NAME "flow_classification"
+#define P4_TABLE_MATCH_FIELDS "{\"priority\":100,\"src_addr\":\"10.1.1.0&255.255.255.0\"}"
+#define P4_TABLE_ACTION "set_tc"
+#define P4_TABLE_ACTION_PARAMETERS "{\"tc\":\"4\"}"
+
+sai_object_id_t p4ext_entry_id = {};
+uint8_t num_attrs = 4;
+sai_attribute_id_t p4ext_entry_attrs[num_attrs] = {};
+
+p4ext_entry_attrs[0].id = (sai_attr_id_t) SAI_P4EXT_ENTRY_ATTR_TABLE_ID;
+p4ext_entry_attrs[0].value.s8list.count = strlen(P4_TABLE_NAME);
+/** P4 Table Name */
+p4ext_entry_attrs[0].value.s8list.value = (sai_s8_list_t.list)P4_TABLE_NAME;
+
+p4ext_entry_attrs[1].id = (sai_attr_id_t) SAI_P4EXT_ENTRY_ATTR_MATCH_FIELD_ID;
+p4ext_entry_attrs[1].value.s8list.count = strlen(P4_TABLE_MATCH_FIELDS);
+/**
+ * Map of match field names and values represented as a string.
```

Comment:
Map is confusing here. It is actually a list represented as a map. Lets just say "Match field keys and values represented as a string".
Same for later occurences.